src/hooks/useActionAddCol.js

```diff
@@ -0,0 +1,26 @@
+import InitialGridConfig from "../config/theme/initial_grid";
+import { GridCell } from "../util/generate_grid_nodes";
+
+const useActionAddCol = (gridNodes, setGridNodes) => {
+  const handleSelect = () => {
+    let rowCount = InitialGridConfig.ROW_COUNT;
+    let newGridNodes = Array.from(gridNodes);
+
+    if (!newGridNodes.length) {
+      for (let i = 0; i < rowCount; i++) {
+        newGridNodes.push([]);
+      }
+    }
+    for (let i = 0; i < newGridNodes.length; i++) {
+      newGridNodes[i].push(new GridCell());
```

Comment:
So one thing is that when you did `Array.from()` is that it's a shallow copy. Since this is a 2D array, it only copied over the references from the inner, rather than copy the entries from the inner array themselves. To make a copy of the inner array as well, you can instead do this:
`newGridNodes[i] = [...newGridNodes[i], new GridCell()]`